src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/LongPollingTransport.java

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+package com.microsoft.signalr;
+
+import io.reactivex.Completable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+public class LongPollingTransport implements Transport {
+    private OnReceiveCallBack onReceiveCallBack;
+    private TransportOnClosedCallback onClose;
+    private String url;
+    private final HttpClient client;
+    private final HttpClient pollingClient = new DefaultHttpClient(100*1000);
+    private final Map<String, String> headers;
+    private volatile Boolean active;
+    private String pollUrl;
+    private final Logger logger = LoggerFactory.getLogger(LongPollingTransport.class);
+
+    public LongPollingTransport(Map<String, String> headers, HttpClient client) {
+        this.headers = headers;
+        this.client = client;
+    }
+
+    @Override
+    public Completable start(String url) {
+        this.active = true;
+        logger.info("Starting LongPolling transport");
+        this.url = url;
+        pollUrl = url + "&_=" + System.currentTimeMillis();
+        logger.info("Polling {}", pollUrl);
+        HttpRequest request = new HttpRequest();
+        request.addHeaders(headers);
+        HttpResponse response = this.client.get(pollUrl, request).blockingGet();
+        if (response.getStatusCode() != 200){
+            logger.error("Unexpected response code {}", response.getStatusCode());
+            this.active = false;
+            return Completable.error(new Exception("Failed to connect"));
+        } else {
+            this.active = true;
+        }
+
+        new Thread(() -> poll(url)).start();
+
+        return Completable.complete();
+    }
+
+    private Completable poll(String url){
+        while(this.active){
+            // Poll
+            pollUrl = url + "&_=" + System.currentTimeMillis();
+            logger.info("Polling {}", pollUrl);
+            HttpRequest request = new HttpRequest();
+            request.addHeaders(headers);
+            HttpResponse response = this.pollingClient.get(pollUrl).blockingGet();
```

Comment:
No `blockingGet`s

Reply:
We have to block here though for the response.

Reply:
poll returns a Completable, so you shouldn't have to block. Like @BrennanConroy suggested, look at HubConnection.handleNegotiate for how to do this. handleNegotiate returns a Single instead of a Completable, but it's trivial to convert: http://reactivex.io/RxJava/javadoc/rx/Single.html#toCompletable--